include/flang/Optimizer/Dialect/FIROps.td

```diff
@@ -0,0 +1,2844 @@
+    llvm::StringRef fieldName;
+    auto &builder = parser.getBuilder();
+    mlir::Type recty;
+    if (parser.parseOptionalKeyword(&fieldName) ||
+        parser.parseComma() ||
+        parser.parseType(recty))
+      return mlir::failure();
+    result.addAttribute(fieldAttrName(), builder.getStringAttr(fieldName));
+    if (!recty.dyn_cast<RecordType>())
+      return mlir::failure();
+    result.addAttribute(typeAttrName(), mlir::TypeAttr::get(recty));
+    mlir::Type lenType = fir::LenType::get(builder.getContext());
+    if (parser.addTypeToList(lenType, result.types))
+      return mlir::failure();
+    return mlir::success();
+  }];
+
+  let printer = [{
+    p << getOperationName() << ' '
+      << getAttrOfType<mlir::StringAttr>(fieldAttrName()).getValue() << ", "
+      << getAttr(typeAttrName());
+  }];
+
+  let builders = [OpBuilder<
+    "Builder *builder, OperationState &result, StringRef fieldName, Type recTy",
+    [{
+      result.addAttribute(fieldAttrName(), builder->getStringAttr(fieldName));
+      result.addAttribute(typeAttrName(), TypeAttr::get(recTy));
+    }]
+  >];
+
+  let extraClassDeclaration = [{
+    static constexpr llvm::StringRef fieldAttrName() { return "field_id"; }
+    static constexpr llvm::StringRef typeAttrName() { return "on_type"; }
+    mlir::Type getOnType() {
+      return getAttrOfType<TypeAttr>(typeAttrName()).getValue();
+    }
+  }];
+}
+
+// Fortran loops
+
+def ImplicitFirTerminator : SingleBlockImplicitTerminator<"FirEndOp">;
+
+def fir_LoopOp : fir_Op<"loop", [ImplicitFirTerminator]> {
+  let summary = "generalized loop operation";
+  let description = [{
+    Generalized high-level looping construct. This operation is similar to
+    MLIR's `loop.for`. An ordered loop will return the final value of `%i`.
+
+    ```mlir
+      %l = constant 0 : index
+      %u = constant 9 : index
+      fir.loop %i = %l to %u unordered {
+        %x = fir.convert %i : (index) -> i32
+        %v = fir.call @compute(%x) : (i32) -> f32
+        %p = fir.coordinate_of %A, %i : (!fir.ref<f32>, index) -> !fir.ref<f32>
+        fir.store %v to %p : !fir.ref<f32>
+      }
+    ```
+
+    The above example iterates over the interval `[%l, %u]`. The unordered
+    keyword indicates that the iterations can be executed in any order.
+  }];
+
+  let arguments = (ins
+    Index:$lowerBound,
+    Index:$upperBound,
+    Variadic<Index>:$optStep,
+    OptionalAttr<I64Attr>:$constantStep,
+    OptionalAttr<UnitAttr>:$unordered
+  );
+
+  let results = (outs Variadic<Index>:$lastVal);
+
+  let regions = (region SizedRegion<1>:$region);
+
+  let skipDefaultBuilders = 1;
+  let builders = [
+    OpBuilder<"mlir::Builder *builder, OperationState &result,"
+              "mlir::Value lowerBound, mlir::Value upperBound,"
+              "ValueRange step = {}, ArrayRef<NamedAttribute> attributes = {}">
+  ];
+
+  let parser = "return parseLoopOp(parser, result);";
+
+  let printer = [{
+    p << getOperationName()
```

Comment:
For the Fortran use case, no. The programmer may specify an ordering to avoid overflows, for example, This is meant to explicitly enforce the determinism specified by the Fortran programmer, while allowing aggressive optimizations in the default case.